keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,799 @@
+- Reject updates to VolumeSnapshot’s VolumeSnapshotSource
+- Reject updates to VolumeSnapshotContent’s VolumeSnapshotContentSource
+- Reject updates to VolumeSnapshotContent’s volume snapshot ref after binding
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Backwards compatibility
+
+There is a backwards compatibility issue involved when tightening the validation on snapshot objects. Since the feature is already in beta we are committed to more backward compatibility guarantee than alpha.
+
+There are two perspectives.
+- client perspective: API requests that previously succeeded now fail (these are okay)
+  - create: invalid things get rejected rather than persisted
+  - update: invalid transitions now get rejected
+- API server perspective for existing persisted data
+  - do not tighten validation (via schema or webhook) in a way that would prevent a no-op update (this must not be violated)
+  - reason: removing a finalizer is done via update, causes problems with delete
+  - making a previously optional field required in the schema blocks update of previously persisted data that omitted the field (unless the update populates the newly required field, or you specify a schema default)
+
+To tackle the backwards compatibility problem, this KEP proposes the following release process.
+
+Begin with validating webhook only enforcement. The webhook will perform the following validation
+- one release with ratcheting validation using the webhook server
+   - webhook is strict on create
+   - webhook is strict on updates where the existing object passes strict validation
+   - webhook is relaxed on updates where the existing object fails strict validation (allows finalizer removal, status update, deletion, etc)
+ - some reconciliation process in volume snapshot controller (this is the hard part) that ensures invalid data is fixed or removed
+
+Once we are sure no invalid data is persisted, we can switch to CRD schema-enforced validation with validating webhooks for immutability in a subsequent release.
+
+We are unsure of the specifics of the reconciliation process in which the controller will remove or fix invalid objects.
+
+#### Current Controller validation of OneOf semantic
+
+##### Handling [VolumeSnapshot](https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/common-controller/snapshot_controller.go#L192).
```

Comment:
Can you reference from a release branch? Otherwise this line could change when new commit gets merged.